python/tank/templatekey.py

```diff
@@ -256,6 +272,124 @@ def _as_string(self, value):
+        # before calling the base class, then repr will crash since self.name won't have been set
+        # yet.
+        if isinstance(format_spec, basestring) is False:
+            raise TankError("format_spec for <Sgtk TimestampKey %s> is not of type string: %s" %
+                            (name, format_spec.__class__.__name__))
+        self.format_spec = format_spec
+
+        if isinstance(default, basestring):
+            # if the user passes in now or utc, we'll generate the current time as the default time.
+            if default.lower() == "now":
+                self._default_to_utc = False
+                default = self.__get_current_time
+            elif default.lower() == "utc_now":
+                self._default_to_utc = True
+                default = self.__get_current_time
+            # Base class will validate other values using the format specifier.
+        elif not(default is None or isinstance(default, datetime.datetime)):
+            raise TankError("default for <Sgtk TimestampKey %s> is not of type string, "
+                            "datetime.datetime or None: %s" % (name, default.__class__.__name__))
+
+        super(TimestampKey, self).__init__(
+            name,
+            default=default
+        )
+
+    def __get_current_time(self):
+        """
+        Returns the current time as a datetime.datetime instance.
+
+        Do not streamline the code so the __init__ method simply passesd the datetime.datetime.now method,
+        we can't mock datetime.now since it's builtin and will make unit tests more complicated to
+        write.
+
+        :returns: If utc_default, a datetime object representing time in the UTC timezone. Otherwise,
+                  a datetime object representing time in the local timezone. In any case, the tzinfo
+                  member is None.
+        """
+        if self._default_to_utc:
+            return datetime.datetime.utcnow()
+        else:
+            return datetime.datetime.now()
+
+    def validate(self, value):
+        """
+        Test if a value is valid for this key.
+
+        :param value: Value to test.
+
+        :returns: Bool
+        """
+        if isinstance(value, basestring):
+            # If we have a string we have to actually try to convert the string to see it if matches
+            # the expected format.
+            try:
+                datetime.datetime.strptime(value, self.format_spec)
+                return True
+            except ValueError, e:
+                # Bad value, report the error to the client code.
+                self._last_error = "Invalid string: %s" % e.message
+                return False
+        elif not isinstance(value, datetime.datetime):
+            self._last_error = "Invalid type: expecting string or datetime.datetime, not %s" % value.__class__.__name__
+            return False
+        else:
+            return True
```

Comment:
Maybe if I flip the elif and else it will be clearer?
```
elif isinstance(value, datetime.datetime):
    return True
else:
    self._last_error = "Invalid type: expecting string or datetime.datetime, not %s" % value.__class__.__name__
    return False
```